File: benchmarks/strmm/wgblas/strmm.js
import { init, cleanup } from "wgblas";
import { strmm } from "wgblas/strmm";
import { GpuMatrix } from "wgblas/classes/GpuMatrix";
import { randomFloat32Array } from "wgblas/random";
import {
  median,
  printHeader,
  printRow,
  getGpuModel,
  saveResults,
  toColumnMajor,
} from "../../utils/helpers.mjs";

// Two-pass pipeline: triangularize (zero-fill op(A)) then an unmodified
// sgemm pass, plus a seed copy (B -> outBuffer, preserves stride-padding
// gaps). uplo/side/transA/diag/A's lda all scoped as non-effects
// (temp/scope_strmm.mjs); the gemm pass's ldb/layout behavior is sgemm's
// own, unswept here — see ldb.sgemm.js/trans.sgemm.js.
const WARMUP_ITERS = 3;
const BENCH_ITERS = 20;
const SIZES = [32, 64, 128, 256, 512, 1024];

const COLS = ["m", "n", "compute_ms", "compute_GFLOPs", "compute_GBs"];

const powerPreference =
  process.argv[2] === "low-power" ? "low-power" : "high-performance";
const device = await init({ benchmark: true, powerPreference });

const gpuModel = getGpuModel();
const records = [];

printHeader(COLS);

for (const size of SIZES) {
  const m = size, n = size; // square: aOrder (side='left') = m = n
  // column-major: lda/ldb >= rows, matching cuBLAS's native layout
  const lda = m, ldb = m;
  const alpha = 1.0;

  const AGpu = GpuMatrix.from(toColumnMajor(randomFloat32Array(m * lda), m, m), m, m, lda, "column-major");
  const BGpu = GpuMatrix.from(toColumnMajor(randomFloat32Array(m * n), m, n), m, n, ldb, "column-major");

  for (let i = 0; i < WARMUP_ITERS; i++) {
    await strmm(device, "left", "lower", "no-transpose", "non-unit", m, n, alpha, AGpu, lda, BGpu, ldb);
  }

  const times = [];
  for (let i = 0; i < BENCH_ITERS; i++) {
    const { gpuTimeMs } = await strmm(
      device, "left", "lower", "no-transpose", "non-unit", m, n, alpha, AGpu, lda, BGpu, ldb,
    );
    if (Number.isFinite(gpuTimeMs) && gpuTimeMs > 0) times.push(gpuTimeMs);
  }

  AGpu.destroy();
  BGpu.destroy();

  if (times.length === 0) continue;

  const med = median(times);
  const aOrder = m; // side='left'
  // gemm-shaped 2*m*n*aOrder multiply-adds, plus m*n for the alpha scale
  // (no beta term — strmm has no C accumulation).
  const flops = 2 * m * n * aOrder + m * n;
  // triangularize: A read + Adense write (2*aOrder^2); seed copy: B read +
  // outBuffer write (2*m*n); gemm: Adense re-read (aOrder^2) + B read +
  // outBuffer read (beta term) + outBuffer write (3*m*n)
  const bytes = (3 * aOrder * aOrder + 5 * m * n) * 4;
  const gflops = flops / 1e9 / (med / 1e3);
  const gbs = bytes / 1e9 / (med / 1e3);
  printRow(COLS, [m, n, med, gflops, gbs]);
  records.push({ m, n, compute_ms: med, compute_GFLOPs: gflops, compute_GBs: gbs });
}

saveResults("strmm", gpuModel, records, { folder: "strmm" });

cleanup();

File: benchmarks/strmm/cuda/strmm.c
#include <stdio.h>
#include <stdlib.h>
#include "../../utils/helpers.h"

#define WARMUP_ITERS 3
#define BENCH_ITERS  20

int main(void) {
    char gpu_model[256];
    get_gpu_model(gpu_model, sizeof(gpu_model));

    cublasHandle_t handle;
    CUBLAS_CHECK(cublasCreate(&handle));

    // Matches benchmark strmm.js: side=left, uplo=lower, trans=no-transpose,
    // diag=non-unit, tight lda/ldb, square m=n (so A's order equals m).
    // cuBLAS's strmm takes a separate C — pass d_C = d_B for the same
    // in-place semantics wgblas's strmm has (B overwritten).
    int sizes[] = { 32, 64, 128, 256, 512, 1024 };
    int num_sizes = (int)(sizeof(sizes) / sizeof(sizes[0]));

    float med_times[num_sizes];
    float gflops_vals[num_sizes];
    float gbs_vals[num_sizes];

    printf("%-10s  %-10s  %-12s  %-14s  %-12s\n",
           "m", "n", "compute_ms", "compute_GFLOPs", "compute_GBs");
    printf("%-10s  %-10s  %-12s  %-14s  %-12s\n",
           "----------", "----------", "------------", "--------------", "------------");

    const float alpha = 1.0f;

    for (int si = 0; si < num_sizes; si++) {
        int m = sizes[si];
        int n = sizes[si];
        int lda = m, ldb = m;

        float *h_A = random_float_array((size_t)m * lda, -1.0f, 1.0f);
        float *h_B = random_float_array((size_t)m * n, -1.0f, 1.0f);

        float *d_A, *d_B;
        CUDA_CHECK(cudaMalloc((void **)&d_A, (size_t)m * lda * sizeof(float)));
        CUDA_CHECK(cudaMalloc((void **)&d_B, (size_t)m * n * sizeof(float)));
        CUDA_CHECK(cudaMemcpy(d_A, h_A, (size_t)m * lda * sizeof(float), cudaMemcpyHostToDevice));
        CUDA_CHECK(cudaMemcpy(d_B, h_B, (size_t)m * n * sizeof(float), cudaMemcpyHostToDevice));

        cudaEvent_t start, stop;
        CUDA_CHECK(cudaEventCreate(&start));
        CUDA_CHECK(cudaEventCreate(&stop));

        for (int i = 0; i < WARMUP_ITERS; i++) {
            CUBLAS_CHECK(cublasStrmm(handle, CUBLAS_SIDE_LEFT, CUBLAS_FILL_MODE_LOWER, CUBLAS_OP_N, CUBLAS_DIAG_NON_UNIT,
                                      m, n, &alpha, d_A, lda, d_B, ldb, d_B, ldb));
        }
        CUDA_CHECK(cudaDeviceSynchronize());

        float compute_times[BENCH_ITERS];
        for (int i = 0; i < BENCH_ITERS; i++) {
            CUDA_CHECK(cudaEventRecord(start, 0));
            CUBLAS_CHECK(cublasStrmm(handle, CUBLAS_SIDE_LEFT, CUBLAS_FILL_MODE_LOWER, CUBLAS_OP_N, CUBLAS_DIAG_NON_UNIT,
                                      m, n, &alpha, d_A, lda, d_B, ldb, d_B, ldb));
            CUDA_CHECK(cudaEventRecord(stop, 0));
            CUDA_CHECK(cudaEventSynchronize(stop));
            CUDA_CHECK(cudaEventElapsedTime(&compute_times[i], start, stop));
        }

        float med = median(compute_times, BENCH_ITERS);
        int a_order = m; // side=left
        float flops = 2.0f * m * n * a_order + (float)m * n;
        float bytes = (float)((size_t)(3 * a_order * a_order + 5 * m * n)) * 4.0f;
        float gflops = (flops / 1e9f) / (med / 1e3f);
        float gbs = (bytes / 1e9f) / (med / 1e3f);

        med_times[si]   = med;
        gflops_vals[si] = gflops;
        gbs_vals[si]    = gbs;

        printf("%-10d  %-10d  %-12.4f  %-14.4f  %-12.4f\n",
               m, n, med, gflops, gbs);

        CUDA_CHECK(cudaEventDestroy(start));
        CUDA_CHECK(cudaEventDestroy(stop));
        cudaFree(d_A);
        cudaFree(d_B);
        free(h_A);
        free(h_B);
    }

    // write JSON results
    record_field fields[] = {
        { "m", FIELD_INT, sizes },
        { "n", FIELD_INT, sizes },
    };
    save_results(gpu_model, "strmm", "strmm", fields, 2, med_times, gbs_vals, gflops_vals, num_sizes);

    cublasDestroy(handle);
    return 0;
}
